enhancements/update/eus-upgrades-mvp.md

```diff
@@ -0,0 +1,366 @@
+
+These changes will need to be backported to 4.7 prior to 4.10 GA.
+
+#### As the maintainer of a Market Place Operator I may define either maxKubeVersion or maxOCPVersion
+
+We should allow maintainers to define an inclusive version range for their Operator.
+
+#### As an admin I cannot violate defined OLM managed Operator compatibility
+
+If a managed Operator defines a maxKubeVersion or maxOCPVersion which would be
+violated by upgrading to the next minor OLM must set Upgradeable=False and
+enumerate the constraint in the condition message.
+
+Note, this is up for debate as to what the behavior should be in absence of a
+defined maximum version.
+
+#### As an admin I have estimated upgrade duration data scaled by node count
+
+With support of the PerfScale team the Over The Air team will measure and
+analyze upgrade timings using simulated workloads at a defined capacity and
+load. This analysis will be used to identify optimizations and set expectations
+on a per minor version basis as to the duration of the upgrade.
+
+Due to the variability in workload deployment these measurements will be
+*exclusive of Worker MCP rollout*. The simulated workload is intended only to
+ensure that our measurements more closely mimic real world usage, not upgrades
+of clusters with zero workload.
+
+#### As an admin, upgrades of CVO managed components are as fast as safe
+
+In 4.6 and 4.7 we identified a number of DaemonSets including dns, mcd, node-tuned,
+node-exporter which were not critical to workload availability which could rollout
+in a more parallel manner. Those operators were changed to utilize maxUnavailable
+of 10% rather than absolute value of 1, on a cluster of 250 nodes that generally
+reduced rollout time from 80 minutes to 10 minutes. The multus, sdn, ovs DaemonSets
+have not been updated but likely could be now that workload critical processes
+have been moved to the host. Before doing this given the risk we should devise
+a method to canary updates to these DaemonSets. Any other options to speed up
+updates to these DaemonSets which are deployed to all Linux nodes should also be
+considered.
+
+Once these new deployment patterns have been vetted in 4.8 or 4.9 these changes
+should be backported through to 4.6.
+
+#### As an admin, my workloads within a failure domain are rescheduled a minimum number of times
+
+Currently the MCO cycles nodes in a deterministic but unintentional manner. The
+workload on that host which can be relocated scatters to other hosts according
+to scheduling rules. This has been measured and shown that approximately 70% of
+all reschedulable pods are rescheduled more than necessary if they were
+rescheduled in an optimal manner. Optimal rescheduling should be closer to (1 /
+node_count_per_failure_domain) percentage of pods being rescheduled more than
+once during a rolling restart of a given failure domain.
+
+We should make the MCO topology aware so that it upgrades all of the host in one
```

Comment:
The MCO should not be topology-aware, IMHO. Topology is very much platform-specific. Really, what is being asked is that the MCO would have to infer sub-pools based on cloud regions. The understanding of cloud-fault domains is well above the level of the MCO. The MCO would need a platform-agnostic means of enumerating the fault-domain and upgrade priorities.

Reply:
I'm no longer certain this is necessary but I'm going to leave this in until I hear back from the workloads team that this isn't necessary.

Reply:
If this is not necessary would prefer to skip as this feels like a separate enhancement altogether.

Reply:
This is really a rollup of many different enhancements across many teams, each likely to lead to an enhancement of their own if sufficiently complex to require an ehancement. I believe we should try to minimize pod rescheduling and I'm currently asking that the workloads team drive that investigation, we'll see what they come up with. I'll change this just to mention that this *may* require coordination between MCO and Scheduler in some form.